torch/csrc/jit/python/script_init.cpp

```diff
@@ -846,6 +846,23 @@ void initJitScriptBindings(PyObject* module) {
     }
   }
 
+  struct DeepCopyMemoTable {
+    std::shared_ptr<IValue::HashAliasedIValueMap> map;
+  };
+  // NOLINTNEXTLINE(bugprone-unused-raii)
+  py::class_<DeepCopyMemoTable>(m, "DeepCopyMemoTable");
+
+  object_class.def(
+      "__deepcopy__", [](const Object& self, const py::dict& memo) {
+        if (!memo.contains(py::str("__torch_script_memo_table"))) {
+          memo["__torch_script_memo_table"] = DeepCopyMemoTable{
+              std::make_shared<IValue::HashAliasedIValueMap>()};
+        }
+        auto& ivalue_memo =
+            *py::cast<DeepCopyMemoTable>(memo["__torch_script_memo_table"]).map;
+        return Object(IValue(self._ivalue()).deepcopy(ivalue_memo).toObject());
```

Comment:
`self._ivallue()` is not an Object? seems like the IValue->Object thing is redundant.

Reply:
no it's an `intrusive_ptr<ivalue::Object>`, not a `script::Object` :p

Reply:
Does IValue not have a deepcopy method?

Reply:
Yeah, it does. `ivalue::Object` doesn't. I think you're mis-reading the parens, `Object(` wraps the final value in an Object, not the result of `_ivalue()`

Reply:
... and somehow toObject also doesn't return an object but an intrusive ptr to what used to be an object impl? The whole IValue and its members has gotten confusing.